### PR TITLE
Properly compute on 31-63 bits in Targetint.OCaml

### PR DESCRIPTION
**Pull request by Gbury**
Previously, the `Targetint.OCaml` module computed on 32/64 bits, which was obviously incorrect (as noted by the comments). This fixes that (at least partially).

The solution is to use a new functor to derive functions that operate on integers that have 1 bit less than what is given to the functor (the name of the file with the functor, `one_bit_less.ml`, could probably be better).

There still quite a few more things to fix in `targetint.ml`, do we want to fix them now or later ?

For the record, before this PR, max_int was "simplified" to `-1`.

### CODE REVIEW

**Comment by mshinwell** [comment]
Ah yeah, sorry about this.  Ugh.  I suggest looking at the Jane Street libraries for this (I think `Base` has an implementation for int63).

**Comment by mshinwell** [comment]
https://github.com/janestreet/base/blob/master/src/int63_emul.ml

**Comment by mshinwell** [comment]
I think it's worth fixing everything in `Targetint` now that might affect Flambda 2.0 compilation.

File: utils/one_bit_less.ml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+  val and_ : t -> t -> t
+  val or_ : t -> t -> t
+  val xor : t -> t -> t
+
+  val shift_left : t -> int -> t
+  val shift_right : t -> int -> t
+  val shift_right_logical : t -> int -> t
+
+  val max : t -> t -> t
```
Comment (lthls) [inline utils/one_bit_less.ml]: You should include `min`, at the very least because `max_string_length` depends on it

**Comment by Gbury** [comment]
@mshinwell do you really prefer if I use an implementation similar to what is done in base, or is the current implementation (of 31 and 63 bits integers) good enough ?
I'll work on fixing the rest of targetint.

**Comment by mshinwell** [comment]
@Gbury Up to you.  I'm not fussed about what the particular implementation is, but it's important that it is correct, and the Base code will have received a lot of review already.

**Comment by Gbury** [comment]
While I'm fixing the targetint module: the overflow specification of conversions function is very chaotic:
- the following function convert modulo the function's return type integer size (i.e. high-order bits are lost): `of_int`, `to_int`, `of_int64`, `to_int32`
- the following functions raise when the argument is outside the function's return type size : `of_int_exn`, `of_string`
- finally the `of_float` function has undefined behaviour if the converted integer is outside the range of target integers.

That makes 1) quite hard to find adequate specification for the corresponding `Targetint.OCaml` conversion functions, 2) very tricky to reason about conversions when using the module because the name of the conversion function doesn't help in figuring out the specification.

For the conversions in the `Targetint.OCaml` module, do we want to mirror the specification of the `Targetint` conversion functions, or do we first try and normalize the overflow specifications ?

**Comment by Gbury** [comment]
I pushed a new version, with added checks on the higher bit for conversions. Currently, the checks raise an exception if the conversion from 32 to 31 or 64 to 63 bits is not exact. This leads to a somewhat strange behaviour for `of_int64` on 32 bit plaforms, where the argument is first converted modulo the range of 32-bits integer, and then the higher order bit is checked to raise en error? We probably want to fix that (by deciding whether it should wrap-around modulo 31 bits, or raise an error if it's outside the range).

Additionally, I factored out the signature used in the implementation (this changes the `targetint.ml` file only, not its interface), and completed the documentation of `targetint.mli`.

@lthls  has reviewed the implementation of the `one_bit_less` functor, I guess one more look at it would be good to ensure there is no error, so if either @mshinwell or @chambart  could take a look, it could be good.

**Comment by mshinwell** [comment]
Please ask @chambart if he could look at this, thanks.  By the way, `one_bit_fewer`, please...

Relating to the functions' behaviours, there are really two kinds of uses:
1. To perform computations that would alternatively have been done at runtime on the target
2. Other miscellaneous conversions within the Flambda code.

Category 1 obviously needs to have the same semantics as the target.

Category 2 probably needs strict error checking.  For example, if we end up with the size of a block that cannot be represented in a `Targetint.OCaml.t`, that should be a fatal error.  A lot of these will arise from the "bad" cross-compilation scenarios (e.g. 32-bit host, 64-bit target), which we may end up not supporting, but should have errors in the Flambda code to catch as much as possible.  We can't easily catch all the cases though without having checks on the sizes of various data structures (e.g. not trying to create arrays that are too big).

**Comment by Gbury** [comment]
This should be good to go after a quick review.
Integration into the `Immediate` module with strongly checked conversions can be done later.

**Comment by mshinwell** [comment]
Please leave a CR so we don't forget about that.

**Comment by Gbury** [comment]
There are CR at the begining of the implementation of each `OCaml` module (e.g. https://github.com/ocaml-flambda/ocaml/pull/32/files#diff-406ab9e39abf0767180065fec9e6fd5fR135 )

**Pull request merged**